File: 189-rotate-array/rotate-array.c
void rotateArray(int *a, int s, int e){
    while (s < e){
        int temp = a[s];
        a[s] = a[e];
        a[e] = temp;
        s++;
        e--;
    }
}

void rotate(int* nums, int numsSize, int k) {
    k = k % numsSize;
    rotateArray(nums, 0, numsSize-1);
    rotateArray(nums, 0, k-1);
    rotateArray(nums, k, numsSize-1);
}

File: 189-rotate-array/rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k % n;
        rotates(0, n - 1, nums);
        rotates(0, k - 1, nums);
        rotates(k, n - 1, nums);
    }

    void rotates(int i, int j, int[] a) {
        while (i < j) {
            int temp = a[i];
            a[i] = a[j];
            a[j] = temp;
            i++;
            j--;
        }
    }
}